source/_posts/2016-09-30-spark-streaming-on-yarn.markdown

```diff
@@ -0,0 +1,324 @@
+    [0-9]*.jvm.pools.G1-Survivor-Space.used
+    [0-9]*.jvm.pools.G1-Eden-Space.used
+
+How much time is spent on GC on Spark executors.
+
+    [0-9]*.jvm.G1-Old-Generation.time
+    [0-9]*.jvm.G1-Young-Generation.time
+
+
+When you configure first Grafana dashboard for Spark job, perhaps the first question will emerge: 
+
+> How to configure Graphite query when metrics for every Spark application run are reported under its own application id?
+
+For driver metrics use wildcard ```.*(application_[0-9]+).*``` and ```aliasSub``` Graphite function to present 'application id' as graph legend:
+
+    aliasSub(stats.analytics.$job_name.*.prod.$dc.*.driver.jvm.heap.used, ".*(application_[0-9]+).*", "heap: \1")
+    
+For executor metrics again use wildcard ```.*(application_[0-9]+).*```, ```groupByNode``` Graphite function to sum metrics from all Spark executors and ```aliasSub``` Graphite function to present 'application id' as graph legend:
+
+    aliasSub(groupByNode(stats.analytics.$job_name.*.prod.$dc.*.[0-9]*.jvm.heap.used, 6, "sumSeries"), "(.*)", "heap: \1")
+
+Finally Grafana dashboard for Spark Job might look like:
+
+{% img /images/blog/spark_job_monitoring.png [spark job monitoring] %}
+
+If Spark application is restarted frequently, metrics for old, already finished runs should be deleted from Graphite.
+Because Graphite does not compact inactive metrics, old metrics slow down Graphite itself and Grafana queries.
+
+## Graceful stop
+
+The last puzzle element is how to stop Spark Streaming application deployed on YARN in a graceful way.
+The standard method for stopping (or rather killing) YARN application is using a command ```yarn application -kill [applicationId]```.
+And this command stops the Spark Streaming application but the application might be killed in the middle of the batch.
+So if the job reads data from Kafka, save processing results on HDFS and finally commit Kafka offsets
+you should expect duplicated data on HDFS when job was stopped just before committing offsets.
+
+The first attempt to solve graceful shutdown issue was to call Spark streaming context stop method in shutdown hook.
+
+``` scala
+sys.addShutdownHook {
+    streamingContext.stop(stopSparkContext = true, stopGracefully = true)
+}
+```
+
+Disappointingly a shutdown hook is called too late to finish started batch and Spark application is killed almost immediately.
+Moreover there is no guarantee that a shutdown hook will be called by JVM at all.
+
+At the time of writing this blog post the only confirmed way to shutdown gracefully Spark Streaming application on YARN
+is to notify somehow the application about planned shutdown, and then stop streaming context programmatically (but not from shutdown hook).
+Command ```yarn application -kill``` should be used only as a last resort if notified application did not stop after defined timeout.
+
+The application can be notified about planned shutdown using marker file on HDFS (the easiest way), 
```

Comment:
Marker file could be stored under path like /var/run/spark_application_name. Marker file works similiar to PID file in UNIX based systems.